packages/common/src/AncillaryDataParser.ts

```diff
@@ -0,0 +1,255 @@
+import type Web3 from "web3";
+import dotenv from "dotenv";
+dotenv.config();
+
+/**
+ * @title Parse ancillary data.
+ * @notice Ancillary data parser implementation following guidelines at:
+ * https://docs.google.com/document/d/1zhKKjgY1BupBGPPrY_WOJvui0B6DMcd-xDR8-9-SPDw/edit
+ * @param {Web3} Web3 instance for accessing important utility functions
+ * @param {String} hex string representation of ancillaryData
+ * @return {Object} parsed ancillary data object.
+ */
+export function parseAncillaryData(web3: Web3, ancillaryData: string): any {
+  // Some requesting contracts set the synthetic token address as ancillary data, so try to parse it first:
+  if (web3.utils.isAddress(ancillaryData)) return { address: ancillaryData };
+  let ancillaryString;
+  try {
+    ancillaryString = web3.utils.hexToUtf8(ancillaryData);
+  } catch (err) {
+    throw "Cannot parse ancillary data bytes to UTF-8!";
+  }
+  return parseAncillaryString(ancillaryString);
+}
+
+// Parses ancillary data string to object.
+function parseAncillaryString(ancillaryString: string): any {
+  const stringObject: any = [];
+  const ancillaryObject: any = {};
+  ancillaryString.split("").forEach((character) => {
+    stringObject.push({ character: character, escape: false, skip: false });
+  });
+  markEscapes(stringObject);
+  const keyValues = splitKeyValues(stringObject);
+  keyValues.forEach((keyValue: any) => {
+    const [key, value] = parseKeyValue(keyValue);
+    ancillaryObject[key] = value;
+  });
+  return ancillaryObject;
+}
+
+// Escapes double quoted keys/values and values enclosed in curly/square brackets.
+function markEscapes(stringObject: any) {
+  stringObject.forEach((charObject: any, openIndex: number, stringObject: any) => {
```

Comment:
You don't _always_ have to provide types for these types of functions. Often times, typescript can infer the types without you having to explicitly specify them. Its ability to do this is not 100%, so you won't _always_ be able to rely on this.

In this case, assuming stringObject has the correct type, it should infer (and you can check this by hovering over in VSCode):
````ts
(charObject: { character: string; escape: boolean; skip: false }, openIndex: number, stringObject: { character: string; escape: boolean; skip: false }[]) => {
```